src/taskgraph/parameters.py

```diff
@@ -195,13 +208,14 @@ def format_spec(spec):
     @staticmethod
     def _fill_defaults(repo_root=None, **kwargs):
         defaults = {}
-        for fn in defaults_functions:
-            defaults.update(fn(repo_root))
-
-        for name, default in defaults.items():
-            if name not in kwargs:
-                kwargs[name] = default
-        return kwargs
+        if repo_root != "SOURCE":
```

Comment:
Hm, I just noticed this line and I'm a bit confused now. If this is set to `SOURCE`, then we'll never call that `_get_defaults()` function in the first place. So are all the changes in there even necessary?

And come to think of it, could the calling code in `mozilla-central` just monkeypatch the `defaults_functions` global variable in this module? I think it might actually be possible to accomplish what you need here without any changes to upstream Taskgraph at all.

Reply:
Let me try to clearify. If repo_root == "SOURCE", that means the root source directory of firefox you are running `mach` in, is like a firefox-115.0.3.tar.xz release source, not a Mercurial or Git checkout. The idea with the patch for mozilla-central is to be able to run `mach build` in a non-cvs folder. (sorry if that was obvious :))

Correct me if im wrong, but will we not always call the `_get_defaults()` function on this row:
`defaults_functions = [_get_defaults]`

However, if we manage to make magic with the `_get_defaults()` without touching upstream.
We still need the changes in upstream taskgraph to `vcs.py` and `generator.py` for `mach` to recognize the folder its run from, if its not a git or mercurial checkout.

Reply:
> Correct me if im wrong, but will we not always call the _get_defaults() function on this row:
defaults_functions = [_get_defaults]

No this isn't calling the function, it's just storing it for later. It isn't invoked until line 213 (`fn(repo_root)`). So all the changes in `_get_defaults` are not necessary (I think).

> We still need the changes in upstream taskgraph to vcs.py and generator.py for mach to recognize the folder its run from, if its not a git or mercurial checkout.

I think they might be unnesseary as well, but I'll start some new comment threads, one for each.

Reply:
this `if` does not seem like its needed either